Update optimizers to allow for custom iterations for constrained fits

kratsg opened the pull request:
# Description

This PR is to help get ready for #285 by making it more manageable. Update optimizers to allow for constrained fits with a configurable number of iterations and handle exceptions better/more appropriately.

This is a much smaller PR.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

Update optimizers to allow for constrained fits with a configurable number of iterations.

lukasheinrich commented on pyhf/optimize/opt_tflow.py:
```diff
@@ -35,18 +53,35 @@ def constrained_bestfit(self, objective, constrained_mu, data, pdf, init_pars, p
         pars = self.tb.concatenate([nuis_cat[:0],poi_par,nuis_cat[0:]])
         objective = objective(pars,data,pdf)
-        hessian   = tf.hessians(objective, nuis_cat)[0]
+        hessian   = tf.hessians(objective, nuis_cat)[0]+1e-10
```
I'm not sure we really want to to this "dark" modification to the hessian that nobody knows is happening. I'd rather try to understand why the fit fails (if we can't figure it out, we should at least print a warning once that we are modifying the hessian). For now, I'd say we should remove this. It was just part of my testing

lukasheinrich commented on pyhf/optimize/opt_tflow.py:
```diff
@@ -11,18 +19,28 @@ def unconstrained_bestfit(self, objective, data, pdf, init_pars, par_bounds):
+                    gradient,
+                    hessian,
+                ], feed_dict={best_fit: best_fit})
+                raise
```
this will probably decrease coverage since we are adding lots of lines that we are not hitting in the tests. not sure how to cover it we would need to perhaps have a `validate_hessian` function that we could test stand-alone?

lukasheinrich commented on pyhf/optimize/opt_tflow.py:
```diff
@@ -35,18 +53,35 @@ def constrained_bestfit(self, objective, constrained_mu, data, pdf, init_pars, p
+                    self.tb.tolist(g),
+                    self.tb.tolist(h),
+                ))
+                raise
```
dito, here

lukasheinrich commented:
thanks @kratsg for continuing to port this stuff over, I added some comments.. if we are not sure how to deal with them, I think it's fine if this PR just adds self.maxiter etc to the optimizers and leaves out the e.g. exception handling and hessian modifications

kratsg commented:
> thanks @kratsg for continuing to port this stuff over, I added some comments.. if we are not sure how to deal with them, I think it's fine if this PR just adds self.maxiter etc to the optimizers and leaves out the e.g. exception handling and hessian modifications

I've removed the exception handling and hessian modifications as I agree that it's probably not needed for this, but was part of the local testing for right now. Until we can figure out why it happens so often in the optimization, etc...
